## Importing Dependecies and packages

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [7]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/deniscornelia123/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [10]:
#printing the stopwords in english
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

## Data Pre-processing

In [11]:
#loading the dataset
news_dataset = pd.read_csv('data/train.csv')

In [12]:
news_dataset.shape

(20800, 5)

In [13]:
news_dataset.head()

id                                              title              author  \
0   0  House Dem Aide: We Didn’t Even See Comey’s Let...       Darrell Lucus   
1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...     Daniel J. Flynn   
2   2                  Why the Truth Might Get You Fired  Consortiumnews.com   
3   3  15 Civilians Killed In Single US Airstrike Hav...     Jessica Purkiss   
4   4  Iranian woman jailed for fictional unpublished...      Howard Portnoy   

                                                text  label  
0  House Dem Aide: We Didn’t Even See Comey’s Let...      1  
1  Ever get the feeling your life circles the rou...      0  
2  Why the Truth Might Get You Fired October 29, ...      1  
3  Videos 15 Civilians Killed In Single US Airstr...      1  
4  Print \nAn Iranian woman has been sentenced to...      1

In [14]:
#count the number of missing values
news_dataset.isnull().sum()

id           0
title      558
author    1957
text        39
label        0
dtype: int64

In [15]:
#replacing the null values with empty string
news_dataset=news_dataset.fillna('')

In [16]:
#merging the author name and news title
news_dataset['content'] = news_dataset['author'] + ''+news_dataset['title']

In [17]:
news_dataset

id                                              title  \
0          0  House Dem Aide: We Didn’t Even See Comey’s Let...   
1          1  FLYNN: Hillary Clinton, Big Woman on Campus - ...   
2          2                  Why the Truth Might Get You Fired   
3          3  15 Civilians Killed In Single US Airstrike Hav...   
4          4  Iranian woman jailed for fictional unpublished...   
...      ...                                                ...   
20795  20795  Rapper T.I.: Trump a ’Poster Child For White S...   
20796  20796  N.F.L. Playoffs: Schedule, Matchups and Odds -...   
20797  20797  Macy’s Is Said to Receive Takeover Approach by...   
20798  20798  NATO, Russia To Hold Parallel Exercises In Bal...   
20799  20799                          What Keeps the F-35 Alive   

                                          author  \
0                                  Darrell Lucus   
1                                Daniel J. Flynn   
2                             Consortiumnews.com   
3                                Jessica Purkiss   
4                                 Howard Portnoy   
...                                          ...   
20795                              Jerome Hudson   
20796                           Benjamin Hoffman   
20797  Michael J. de la Merced and Rachel Abrams   
20798                                Alex Ansary   
20799                              David Swanson   

                                                    text  label  \
0      House Dem Aide: We Didn’t Even See Comey’s Let...      1   
1      Ever get the feeling your life circles the rou...      0   
2      Why the Truth Might Get You Fired October 29, ...      1   
3      Videos 15 Civilians Killed In Single US Airstr...      1   
4      Print \nAn Iranian woman has been sentenced to...      1   
...                                                  ...    ...   
20795  Rapper T. I. unloaded on black celebrities who...      0   
20796  When the Green Bay Packers lost to the Washing...      0   
20797  The Macy’s of today grew from the union of sev...      0   
20798  NATO, Russia To Hold Parallel Exercises In Bal...      1   
20799    David Swanson is an author, activist, journa...      1   

                                                 content  
0      Darrell LucusHouse Dem Aide: We Didn’t Even Se...  
1      Daniel J. FlynnFLYNN: Hillary Clinton, Big Wom...  
2      Consortiumnews.comWhy the Truth Might Get You ...  
3      Jessica Purkiss15 Civilians Killed In Single U...  
4      Howard PortnoyIranian woman jailed for fiction...  
...                                                  ...  
20795  Jerome HudsonRapper T.I.: Trump a ’Poster Chil...  
20796  Benjamin HoffmanN.F.L. Playoffs: Schedule, Mat...  
20797  Michael J. de la Merced and Rachel AbramsMacy’...  
20798  Alex AnsaryNATO, Russia To Hold Parallel Exerc...  
20799             David SwansonWhat Keeps the F-35 Alive  

[20800 rows x 6 columns]

In [18]:
#separating data and labels
X= news_dataset.drop('label', axis=1)
y = news_dataset['label']

In [20]:
X.shape

(20800, 5)

In [21]:
y.shape

(20800,)

### Stemming
Stemming is a process of reducing a word to its root word

In [22]:
port_stem =PorterStemmer()

In [29]:
def stemming(content):
    stemmed_content = re.sub('[^a-zA-Z]',' ', content)
    stemmed_content = stemmed_content.lower()
    stemmed_content = stemmed_content.split()
    stemmed_content = [port_stem.stem(word) for word in stemmed_content if not word in stopwords.words('english')]
    stemmed_content = ' '.join(stemmed_content)
    return stemmed_content

In [30]:
news_dataset['content']

0        Darrell LucusHouse Dem Aide: We Didn’t Even Se...
1        Daniel J. FlynnFLYNN: Hillary Clinton, Big Wom...
2        Consortiumnews.comWhy the Truth Might Get You ...
3        Jessica Purkiss15 Civilians Killed In Single U...
4        Howard PortnoyIranian woman jailed for fiction...
                               ...                        
20795    Jerome HudsonRapper T.I.: Trump a ’Poster Chil...
20796    Benjamin HoffmanN.F.L. Playoffs: Schedule, Mat...
20797    Michael J. de la Merced and Rachel AbramsMacy’...
20798    Alex AnsaryNATO, Russia To Hold Parallel Exerc...
20799               David SwansonWhat Keeps the F-35 Alive
Name: content, Length: 20800, dtype: object

In [31]:
news_dataset['content'] = news_dataset['content'].apply(stemming)

In [32]:
news_dataset['content']

0        darrel lucushous dem aid even see comey letter...
1        daniel j flynnflynn hillari clinton big woman ...
2                consortiumnew comwhi truth might get fire
3        jessica purkiss civilian kill singl us airstri...
4        howard portnoyiranian woman jail fiction unpub...
                               ...                        
20795    jerom hudsonrapp trump poster child white supr...
20796    benjamin hoffmann f l playoff schedul matchup ...
20797    michael j de la merc rachel abramsmaci said re...
20798    alex ansarynato russia hold parallel exercis b...
20799                        david swansonwhat keep f aliv
Name: content, Length: 20800, dtype: object

In [35]:
#separting the data and model
X = news_dataset['content'].values
y = news_dataset['label'].values

In [36]:
print(X)

['darrel lucushous dem aid even see comey letter jason chaffetz tweet'
 'daniel j flynnflynn hillari clinton big woman campu breitbart'
 'consortiumnew comwhi truth might get fire' ...
 'michael j de la merc rachel abramsmaci said receiv takeov approach hudson bay new york time'
 'alex ansarynato russia hold parallel exercis balkan'
 'david swansonwhat keep f aliv']


In [37]:
print(y)

[1 0 1 ... 0 1 1]


In [38]:
#converting textual data to numerical data
vectorizer = TfidfVectorizer()
vectorizer.fit(X)
X = vectorizer.transform(X)

In [39]:
print(X)

  (0, 26340)	0.28088379401596425
  (0, 22724)	0.2552336018069161
  (0, 15019)	0.43006226759639316
  (0, 14555)	0.29177259684200296
  (0, 12782)	0.24619727512767195
  (0, 8022)	0.23133661742488731
  (0, 6273)	0.2839932825877813
  (0, 5969)	0.35488202138141456
  (0, 5006)	0.2472595823572816
  (0, 4211)	0.3625320323150658
  (0, 578)	0.2694167078545385
  (1, 27923)	0.36911845953845024
  (1, 11313)	0.24166773097712638
  (1, 8772)	0.5258635625386451
  (1, 5916)	0.31810058109638056
  (1, 4767)	0.23338756776626793
  (1, 3859)	0.45980466668763476
  (1, 3281)	0.18652439327549428
  (1, 2622)	0.3562953366945267
  (2, 26235)	0.3665032495181434
  (2, 16361)	0.43295215406038445
  (2, 9454)	0.30743020569262086
  (2, 8567)	0.3411947414020896
  (2, 5240)	0.40440534260277944
  (2, 5121)	0.5511414848555652
  :	:
  (20797, 25776)	0.08220218573989037
  (20797, 25319)	0.3119640221826561
  (20797, 22086)	0.24902354987792552
  (20797, 20778)	0.2729578683228216
  (20797, 20493)	0.249994989010826
  (20797, 17505

In [44]:
#splitting the dataset into training and testing set
X_train,X_test,y_train, y_test = train_test_split(X,y,test_size=0.2, stratify=y, random_state=2)

In [45]:
#training the model
model =LogisticRegression()

In [46]:
model.fit(X_train,y_train)

LogisticRegression()

In [48]:
#prediction on training data
X_train_predict = model.predict(X_train)
X_train_accuracy = accuracy_score(X_train_predict, y_train)

In [49]:
print(X_train_accuracy)

0.9719951923076923


In [50]:
X_test_predict = model.predict(X_test)
X_test_accuracy = accuracy_score(X_test_predict, y_test)

In [52]:
print("the accuracy score for test data: ", X_test_accuracy)

the accuracy score for test data:  0.9548076923076924


Making a predictive system

In [59]:
X_new = X_test[5]

prediction = model.predict(X_new)
print(prediction)
if(prediction[0]==0):
    print('the news is real')
else:
    print('the news is fake')

[1]
the news is fake


In [55]:
prediction

array([0])

In [60]:
print(y_test[5])

1


## prediction with knn

In [61]:
from sklearn.neighbors import KNeighborsClassifier

In [64]:
model2 = KNeighborsClassifier()
model2.fit(X_train, y_train)
X_train_predict2 = model2.predict(X_train)

In [67]:
model2_accuracy = accuracy_score(X_train_predict2,y_train)
model2_accuracy

0.5168269230769231

In [68]:
X_test_predict2 = model2.predict(X_test)
X_test_accuracy2 = accuracy_score(X_test_predict2, y_test)

In [69]:
print('Accuracy score of the test_data using KNN', X_test_accuracy2)

Accuracy score of the test_data using KNN 0.510576923076923


In [76]:
X_new2 = X_test[5]
model2.predict(X_new2)

array([1])

In [77]:
print(y_test[5])

1


The model predicted better with LogisticRegression with 97% accuracy